### ISSUE TITLE
 [Possible bug?] items.sqlite3-wal might be growing too large

### ISSUE THREAD

**Issue by lhorace** [abraunegg/onedrive#309]
### Bug Report Details ###
**Describe the bug**
Hello,

I have a relatively small OneDrive folder, see [1]. While I was copying over my current onedrive config over to a new partition, the operation lasted longer than expected. Divelging in, I find items.sqlite3-wal file to be occuping 141M of disk space. Is this normal behavior? I’m bit cautions about the size of that file as it reside a relatively small SSD /home partition?

The little information I could find about that file, is that it's part of SQlite `Write-Adhead Logging` feature. See [2], [3], and [4] for additional context. At first glance in reference to [2], this could be a problem if SQlite `Write-Adhead Logging` is not handled properly.

The version of onedrive at the time was `onedrive v2.2.1-7-g91aaf91`

1. `29M home/user/Cloud/OneDrive`
2. https://sourceforge.net/p/liferea/mailman/message/29327620/
3. https://stackoverflow.com/questions/20969996/is-it-safe-to-delete-sqlites-wal-file
4. https://stackoverflow.com/questions/20969996/is-it-safe-to-delete-sqlites-wal-file

**Application and Operating System Details:**
- OS: ArchLinux `Linux linux-desktop 4.19.11-1-ck-zen #1 SMP PREEMPT Thu Dec 20 05:03:12 EST 2018 x86_64 GNU/Linux`
- Are you using a headless system (no gui) or with a gui installed? with gui
- Application version: `onedrive v2.2.3-1-g1060d85`
- OneDrive Account Type: Personal
- DMD or LDC compiler version: `dmd 1:2.083.1-1`

**To Reproduce**
Steps to reproduce the behavior if not causing an application crash:
Run onedrive overtime.

**Complete Verbose Log Output**
N/A

Application Log Output:
N/A

**Screenshots**
N/A

**Additional context**
If there are any queries, please ask.

### Bug Report Checklist ###
- [ ] Detailed description
- [ ] Reproduction steps (if applicable)
- [ ] Verbose Log Output

**Comment by abraunegg** [comment]
@lhorace 
Were you running the client as a 'service' via systemd / initd or 'manually'? Interested to know how the timestamps of all the 'items.sqlite3' files in your config directory & sizes.

If running as a service - how 'long' has the 'service' been running for?

**Comment by lhorace** [comment]
1. I always run it as a `systemd user` service.
2.
```
➜  onedrive ll
total 141M
-rw-r--r-- 1 nouser nouser  349 Dec 14 09:31 config
-rw-r--r-- 1 nouser nouser 380K Dec 21 11:08 items.sqlite3
-rw-r--r-- 1 nouser nouser  32K Dec 21 15:43 items.sqlite3-shm
-rw-r--r-- 1 nouser nouser 140M Dec 21 15:43 items.sqlite3-wal
-rw-r--r-- 1 nouser nouser  409 Dec 21 15:25 refresh_token
```
3. Onedrive was installed on `[2018-12-14 09:23] [PACMAN] Running '/usr/bin/pacman --config /etc/pacman.conf -U /home/user/build/onedrive-abraunegg-git/onedrive-abraunegg-git-1:2.2.1.r7.g91aaf91-1-x86_64.pkg.tar.xz'
` and enable it (started it) on `Dec 16 04:58:33 linux-desktop systemd[723]: Started OneDrive Free Client.`

**Comment by abraunegg** [comment]
So after restart `items.sqlite3-wal` dropped from 140Mb -> ~4Mb .. this would seem normal.

Will have a read of the links as well as this is certainly something related to how the application was originally written to use sqlite. To date the database fixes that have been made were all around ensuring all data was correctly being entered into the database, fixing a lot of the 'parent cannot be found errors'

**Comment by lhorace** [comment]
> So after restart items.sqlite3-wal dropped from 140Mb -> ~4Mb .. this would seem normal.

No, there are at two different locations, full path for clarifty:

**Old:**
```
➜  ~ ll /mnt/disks/hitachi/p2_btrfs/home/user/.config/onedrive
total 141M
-rw-r--r-- 1 nouser nouser  349 Dec 14 09:31 config
-rw-r--r-- 1 nouser nouser 380K Dec 21 11:08 items.sqlite3
-rw-r--r-- 1 nouser nouser  32K Dec 21 15:43 items.sqlite3-shm
-rw-r--r-- 1 nouser nouser 140M Dec 21 15:43 items.sqlite3-wal
-rw-r--r-- 1 nouser nouser  409 Dec 21 15:25 refresh_token
```
**New:**
```
➜  ~ ll /home/nouser/.config/onedrive 
total 4.4M
-rw-r--r-- 1 nouser nouser  349 Dec 21 16:46 config
-rw-r--r-- 1 nouser nouser 376K Dec 21 18:12 items.sqlite3
-rw-r--r-- 1 nouser nouser  32K Dec 21 18:14 items.sqlite3-shm
-rw-r--r-- 1 nouser nouser 4.0M Dec 21 18:14 items.sqlite3-wal
-rw-r--r-- 1 nouser nouser  409 Dec 21 17:28 refresh_token
```
I can easily boot into the old ArchLinux and verify the behavior.

**Comment by abraunegg** [comment]
OK .. if you can clarify & confirm behavior with latest master after running for a at least a week without the service restarting, then we can determine if we have an actual problem here. In the meantime I will look at the sqlite code side of things and see if there is anything obvious.

**Comment by norbusan** [comment]
Interesting. I see the same growing .WAL behavior. Reading through the sqlite.d and itemdb.d code **I** don't see anything blaring. The first link hints at adding `sqlite3_reset` should fix this, but also AFAIS every statement is executed via `sqlite.d:exec` which calls `sqlite.d:reset` which calls `sqlite3_reset`, so this should not be a problem. But well, actually no idea about sqlite whatsoever ...

**Comment by lhorace** [comment]
If I had the time, I could look into it, but my knowledge of SQlite is sparse at best, notwithstanding `dlang`. Just roughly 24 hours in:

`onedrive v2.2.3-1-g1060d85`

```
{18-12-22 12:31}linux-desktop:~ nouser% ll -h .config/onedrive
total 35M
drwxr-xr-x  2 nouser nouser  112 Dec 21 17:28 .
drwxr-xr-x 20 nouser nouser 4.0K Dec 22 02:13 ..
-rw-r--r--  1 nouser nouser  349 Dec 21 16:46 config
-rw-r--r--  1 nouser nouser 376K Dec 22 11:52 items.sqlite3
-rw-r--r--  1 nouser nouser  96K Dec 22 12:31 items.sqlite3-shm
-rw-r--r--  1 nouser nouser  35M Dec 22 12:31 items.sqlite3-wal
-rw-r--r--  1 nouser nouser  409 Dec 22 11:36 refresh_token

```

**Comment by abraunegg** [comment]
> Interesting. I see the same growing .WAL behavior. Reading through the sqlite.d and itemdb.d code **I** don't see anything blaring. The first link hints at adding `sqlite3_reset` should fix this, but also AFAIS every statement is executed via `sqlite.d:exec` which calls `sqlite.d:reset` which calls `sqlite3_reset`, so this should not be a problem. But well, actually no idea about sqlite whatsoever ...

Likewise .. I cannot see anything 'glaring'.

The only thing I can think of here is 'something' in the code is not calling / cleaning up the database connection use correctly - or potentially how it is being 'opened' in monitor mode & kept open is contributing to the issue.

In monitor mode, we open the database before deciding to loop into 'monitor'. Potentially it may be better to 'open' the database on each loop start? Why have the database open during 'monitor_interval' - the period between sync's

**Comment by norbusan** [comment]
Hmm, for changes coming from the cloud that would be fine, but the file system monitor could kick of for every change, and how would we want to deal with that?

I don't think that the opening of the connection is the problem (but of course, closing it would clear the back log), but un-finished/released prepared statements, that is how far I read the links sent. But no decent idea. If I have time in the next week I might write some testing code into the sqlite.d

**Comment by lhorace** [comment]
I think, at least for now, properly closing the DB should be the goal for now, and consider this issue as part of the rewrite? Anyway, here is some useful information I found:

> 2.1. Checkpointing
Of course, one wants to eventually transfer all the transactions that are appended in the WAL file back into the original database. Moving the WAL file transactions back into the database is called a "checkpoint".

> Another way to think about the difference between rollback and write-ahead log is that in the rollback-journal approach, there are two primitive operations, reading and writing, whereas with a write-ahead log there are now three primitive operations: reading, writing, and checkpointing.

> By default, SQLite does a checkpoint automatically when the WAL file reaches a threshold size of 1000 pages. (The SQLITE_DEFAULT_WAL_AUTOCHECKPOINT compile-time option can be used to specify a different default.) Applications using WAL do not have to do anything in order to for these checkpoints to occur. But if they want to, applications can adjust the automatic checkpoint threshold. Or they can turn off the automatic checkpoints and run checkpoints during idle moments or in a separate thread or process.

**Comment by norbusan** [comment]
Yes, but as it seems the automatic checkpointing does not work, and we probably run into the same errors as listed in the link https://sourceforge.net/p/liferea/mailman/message/29327620/ that a manual check point does not work.

I played around a bit and was surprised that a `systemctl -u stop onedrive` followed by `start` did **not** clear the wal file at all ... which is not really good.

**Comment by abraunegg** [comment]
> I played around a bit and was surprised that a `systemctl -u stop onedrive` followed by `start` did **not** clear the wal file at all ... which is not really good.

OK .. that does not seem right .. it should be clearing itself up on stop/start operations. This would mean that it is not 'closing' something properly somewhere.

> Is there any critical information being stored in the SQLite DB? Can it just be regenerated on corruption?

The DB is essentially a 'cache' of what has been transferred, so no issues in being regenerated on corruption. In the 'skilion' version, what would happen is files would then get re-uploaded. In this branch, if a file is not in the DB, it checks first if the file is on OneDrive & if cound, compares this to the local file. If upload is needed - then it is uploaded, otherwise the DB is just updated.

**Comment by norbusan** [comment]
I added code to manually create a checkpoint, but
```
Dec 23 17:37:40 bulldog onedrive[979]: DEBUG trying to create checkpoint
Dec 23 17:37:40 bulldog onedrive[979]: Checkpoint return: 6
```
So it seems it is the same problem ...

**Comment by norbusan** [comment]
Checking with the headers,6 = `SQLITE_LOCKED` which means
> The SQLITE_LOCKED result code differs from SQLITE_BUSY in that SQLITE_LOCKED indicates a conflict on the same database connection (or on a connection with a shared cache) whereas SQLITE_BUSY indicates a conflict with a different database connection, probably in a different process.

**Comment by norbusan** [comment]
Ok, I guess I know what is the problem, but please consider this a guess of an sqlite beginner

- the `ItemDatabase` uses several statements `insertItemStmt` etc which are `prepared` and the again and again `bind` to values and executed
- these statements are never `finalized`
- as long as a statement is not finalized, it keeps a write lock on the database

I confirmed this by calling on each monitor mode sync round a routine that prints out all the prepared and open statements:
```
        void dump_open_statements()
        {
                log.log("Dumpint open statements: \n");
                auto p = sqlite3_next_stmt(pDb, null);
                while (p != null) {
                        log.log (" - " ~ ifromStringz(sqlite3_sql(p)) ~ "\n");
                        p = sqlite3_next_stmt(pDb, p);
                }
        }
```
which gives exactly the following statements from `itemdb.d`

- `insertItemStmt`
- `updateItemStmt`
- `selectItemByIdStmt`
- `selectItemByParentIdStmt`
- `deleteItemByIdStmt`

These statements are reused again and again with `reset`, `bind`, `exec`, but since they are never actually finalized, they keep the database locked, and the checkpointing cannot occur.

I have now pushed a branch https://github.com/abraunegg/onedrive/tree/norbert/sqlite-wal that makes the following changes:

- converts the sql statements to simple strings
- use `db.prepare` on any invocation, the destructor of the Statement class will call finalize
- add manual checkpoints and dumping of open statements

With this I get
```
Syncing changes from OneDrive ...
DEBUG trying to create checkpoint
Checkpoint return: 0
Dumpint open statements: 

Syncing changes from OneDrive ...
```
so it seems the checkpointing works and there are no open statements.

BUT: my .wal file is still the same big size and is not reset to 0 size ... strange ...

**Comment by norbusan** [comment]
Ok, and here are (hopefully) the last information necessary:

- running `onedrive --synchronize` properly cleans up the wal file
- while running in monitor mode, exiting is done with Ctrl-C which does **not** properly shut down the database connection, so the -wal file remains (**and** is important and necessary!)
- running with a clean up -wal file (ie after `--synchronize`) again in `--monitor` makes the wal file reappear but only grow to about 33k size

In the last step I removed the code that automatically creates checkpoints and let sqlite do the automatic checkpointing. With that the wal of course reappears again, but grows only according to the docs
> SQLite will automatically checkpoint whenever a COMMIT occurs that causes the WAL file to be 1000 pages or more in size
I didn't actually check the size but after a few sync runs the size is still at 33k as before.

That means that with the current code in the above branch wal files don't grow arbitrarily anymore.

What still needs to be done is handling Ctrl-C or systemd stop signal handler to shut down the connection properly.

**Comment by norbusan** [comment]
Last comment for today .. I tried to catch Ctrl-C using what I know from C:
```
                        extern(C) @nogc void exitHandler(int value) {
                                printf("Ooohhhh got %d\n", value);
                                exit(0);
                        }
                        signal(SIGINT, &exitHandler);
```
just before going into the monitor infinite loop. That worked, it reacted to Ctrl-C, printed the value, but then segfaults ...

I did put the signal handler right into the main function because I need access to the `itemDb` to close it for proper shutdown, but that might be the source.

If anyone has a nice idea how this can be handled better, please let me know. 1h of searching didn't show up anything really useful.

Thanks

**Comment by lhorace** [comment]
Hey @norbusan, when you get excited, you really put in a lot effort, it's very encouraging! Nevertheless, I looked at your PR, do you have any sample code where you were experimenting with catching signals?

**Comment by lhorace** [comment]
@abraunegg Seems reasonable, although as @norbusan pointed out, one would need to clear out `SQLITE_LOCKED` prior to running a manual `check point`

**Comment by norbusan** [comment]
> Hey @norbusan, when you get excited, you really put in a lot effort, it's very encouraging! Nevertheless, I looked at your PR, do you have any sample code where you were experimenting with catching signals?

Thanks for the flowers, but it is simply that I use onedrive a lot and need to get as many bugs fixed as possible ;-) despite me never having programmed D at all ;-)

Catching signal code I pasted above, here is the full diff
```
--- a/src/main.d
+++ b/src/main.d
@@ -1,8 +1,9 @@
-import core.stdc.stdlib: EXIT_SUCCESS, EXIT_FAILURE;
+import core.stdc.stdlib: EXIT_SUCCESS, EXIT_FAILURE, exit;
 import core.memory, core.time, core.thread;
 import std.getopt, std.file, std.path, std.process, std.stdio, std.conv, std.algorithm.searching, std.string;
 import config, itemdb, monitor, onedrive, selective, sync, util;
 import std.net.curl: CurlException;
+import core.stdc.signal;
 static import log;
 
 int main(string[] args)
@@ -482,6 +483,12 @@ int main(string[] args)
                                        log.logAndNotify("Cannot move item:, ", e.msg);
                                }
                        };
+                       extern(C) @nogc void exitHandler(int value) {
+                               printf("Ooohhhh got %d\n", value);
+                               exit(0);
+                       }
+                       signal(SIGINT, &exitHandler);
+
                        // initialise the monitor class
                        if (cfg.getValue("skip_symlinks") == "true") skipSymlinks = true;
                        if (!downloadOnly) m.init(cfg, verbose, skipSymlinks);
```

**Comment by norbusan** [comment]
> @abraunegg Seems reasonable, although as @norbusan pointed out, one would need to clear out `SQLITE_LOCKED` prior to running a manual `check point`

There is no need for that anymore after my PR https://github.com/abraunegg/onedrive/pull/310 is included. To clean up we could easily add an open/close/open cycle at the beginning of the program.

**Comment by norbusan** [comment]
@lhorace concerning the signal handling code, see the patch above, or look at the branch I just pushed https://github.com/abraunegg/onedrive/tree/norbert/signal-handling
Please contribute/improve/fix/help

**Comment by lhorace** [comment]
I have a few questions that is a bit off-topic, what are opinions about dlang? I want to learn a new language and I was thing Go-lang?

I don't know anything about dlang but the code doesn't seem really hard, the difficulty is probably the syntax.

**Comment by lhorace** [comment]
And the logic.

**Comment by norbusan** [comment]
I don't have an option on D, I never used it till I got somehow interested in this project. I consider all programming languages more or less the same. If you know one, you know most. The thing you have to learn is the syntax and a bit the paradigm, but there are not so many paradigms anyway (declarative, functional, algebraic, ...).

Go and Rust are surely the big shooting stars now. Personally I like Scala a lot.

**Comment by lhorace** [comment]
Thanks for the feedback, I'll give Go-lang a go first (the irony), just figuring out what program I like to create to get me started. Then eventually that experience there could help you guys here. :)

_ends off-topic_
